cura/Machines/MaterialManager.py

```diff
@@ -0,0 +1,307 @@
+
+            default_root_material_id = data_dict.get(self._default_approximate_diameter_for_quality_search)
+            if default_root_material_id is None:
+                default_root_material_id = list(data_dict.values())[0]  # no default diameter present, just take "the" only one
+            for root_material_id in data_dict.values():
+                self._diameter_material_map[root_material_id] = default_root_material_id
+
+        # Map #4
+        #    "machine" -> "variant_name" -> "root material ID" -> specific material InstanceContainer
+        # Construct the "machine" -> "variant" -> "root material ID" -> specific material InstanceContainer
+        self._diameter_machine_variant_material_map = dict()
+        for material_metadata in material_metadata_list:
+            # We don't store empty material in the lookup tables
+            if material_metadata["id"] == "empty_material":
+                continue
+
+            root_material_id = material_metadata["base_file"]
+            definition = material_metadata["definition"]
+            approximate_diameter = material_metadata["approximate_diameter"]
+
+            if approximate_diameter not in self._diameter_machine_variant_material_map:
+                self._diameter_machine_variant_material_map[approximate_diameter] = {}
+
+            machine_variant_material_map = self._diameter_machine_variant_material_map[approximate_diameter]
+            if definition not in machine_variant_material_map:
+                machine_variant_material_map[definition] = MaterialNode()
+
+            machine_node = machine_variant_material_map[definition]
+            variant_name = material_metadata.get("variant_name")
+            if not variant_name:
+                # if there is no variant, this material is for the machine, so put its metadata in the machine node.
+                machine_node.material_map[root_material_id] = MaterialNode(material_metadata)
+            else:
+                # this material is variant-specific, so we save it in a variant-specific node under the
+                # machine-specific node
+                if variant_name not in machine_node.children_map:
+                    machine_node.children_map[variant_name] = MaterialNode()
+
+                variant_node = machine_node.children_map[variant_name]
+                if root_material_id not in variant_node.material_map:
+                    variant_node.material_map[root_material_id] = MaterialNode(material_metadata)
+                else:
+                    # Sanity check: make sure we don't have duplicated variant-specific materials for the same machine
+                    raise RuntimeError("Found duplicate variant name [%s] for machine [%s] in material [%s]" %
+                                       (variant_name, definition, material_metadata["id"]))
+
+        self.materialsUpdated.emit()
+
+    def _updateMaps(self):
+        self.initialize()
```

Comment:
Agree with the performance concern. We also had this concern in the beginning, but the result turned out to be fine. The construction of the tables/maps/trees are extremely fast and there is no noticeable slowness, so we decide to keep it like this for now because it's simple.